azure_functions_worker/shared_memory_data_transfer/shared_memory_manager.py

```diff
@@ -0,0 +1,172 @@
+
+    def is_enabled(self) -> bool:
+        """
+        Whether supported types should be transferred between functions host and
+        the worker using shared memory.
+        """
+        return is_envvar_true(
+            FUNCTIONS_WORKER_SHARED_MEMORY_DATA_TRANSFER_ENABLED)
+
+    def is_supported(self, datum: Datum) -> bool:
+        """
+        Whether the given Datum object can be transferred to the functions host
+        using shared memory.
+        This logic is kept consistent with the host's which can be found in
+        SharedMemoryManager.cs
+        """
+        if datum.type == 'bytes':
+            num_bytes = len(datum.value)
+            if num_bytes >= consts.MIN_BYTES_FOR_SHARED_MEM_TRANSFER and \
+                    num_bytes <= consts.MAX_BYTES_FOR_SHARED_MEM_TRANSFER:
+                return True
+        elif datum.type == 'string':
+            num_bytes = len(datum.value) * consts.SIZE_OF_CHAR_BYTES
+            if num_bytes >= consts.MIN_BYTES_FOR_SHARED_MEM_TRANSFER and \
+                    num_bytes <= consts.MAX_BYTES_FOR_SHARED_MEM_TRANSFER:
+                return True
+        return False
+
+    def put_bytes(self, content: bytes) -> Optional[SharedMemoryMetadata]:
+        """
+        Writes the given bytes into shared memory.
+        Returns metadata about the shared memory region to which the content was
+        written if successful, None otherwise.
+        """
+        if content is None:
+            return None
+        mem_map_name = str(uuid.uuid4())
+        content_length = len(content)
+        shared_mem_map = self._create(mem_map_name, content_length)
+        if shared_mem_map is None:
+            return None
+        num_bytes_written = shared_mem_map.put_bytes(content)
+        if num_bytes_written != content_length:
+            logger.error(
+                f'Cannot write data into shared memory {mem_map_name} '
+                f'({num_bytes_written} != {content_length})')
+            shared_mem_map.dispose()
+            return None
+        self.allocated_mem_maps[mem_map_name] = shared_mem_map
+        return SharedMemoryMetadata(mem_map_name, content_length)
+
+    def put_string(self, content: str) -> Optional[SharedMemoryMetadata]:
+        """
+        Writes the given string into shared memory.
+        Returns the name of the memory map into which the data was written if
+        succesful, None otherwise.
+        """
+        if content is None:
+            return None
+        content_bytes = content.encode('utf-8')
+        return self.put_bytes(content_bytes)
+
+    def get_bytes(self, mem_map_name: str, offset: int, count: int) \
+            -> Optional[bytes]:
+        """
+        Reads data from the given memory map with the provided name, starting at
+        the provided offset and reading a total of count bytes.
+        Returns the data read from shared memory as bytes if successful, None
+        otherwise.
+        """
+        if offset != 0:
```

Comment:
We do not support positive offsets at least?

Reply:
The support is there but right now we don't expect the host to send any non-zero offset messages (as there is no use case) so disallowing it.